lib/saluki-common/src/cache/expiry.rs

```diff
@@ -0,0 +1,251 @@
+            },
+            ExpirationOp::Removed(key) => {
+                // When an entry is removed, we also remove it from the last seen map.
+                self.last_seen.remove(&key);
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct State<K> {
+    inner: Mutex<Inner<K>>,
+    pending_ops: ArrayQueue<ExpirationOp<K>>,
+}
+
+impl<K> State<K>
+where
+    K: Eq + std::hash::Hash,
+{
+    fn new(time_to_idle: Duration) -> Self {
+        Self {
+            inner: Mutex::new(Inner {
+                last_seen: PrehashedHashMap::default(),
+                time_to_idle,
+            }),
+            pending_ops: ArrayQueue::new(128),
+        }
+    }
+
+    fn mark_entry_accessed(&self, key: K) {
+        if let Err(op) = self.pending_ops.push(ExpirationOp::Accessed(key)) {
+            self.process_pending_operations(Some(op));
+        }
+    }
+
+    fn mark_entry_removed(&self, key: K) {
+        if let Err(op) = self.pending_ops.push(ExpirationOp::Removed(key)) {
+            self.process_pending_operations(Some(op));
+        }
+    }
+
+    fn process_pending_operations(&self, pending_op: Option<ExpirationOp<K>>) {
+        let mut inner = self.inner.lock().unwrap();
+        while let Some(op) = self.pending_ops.pop() {
+            inner.process_operation(op);
+        }
+
+        if let Some(pending_op) = pending_op {
+            inner.process_operation(pending_op);
+        }
+    }
+}
+
+/// Expiration management.
+///
+/// This type provides an interface the core expiration logic, allowing the marking of entries when they're accessed as
+/// well as determining when a tracked entry has expired so it can be removed from the cache.
+#[derive(Clone, Debug)]
+pub struct Expiration<K> {
+    state: Option<Arc<State<K>>>,
+}
+
+impl<K> Expiration<K>
+where
+    K: Eq + std::hash::Hash,
+{
+    fn disabled() -> Self {
+        Self { state: None }
+    }
+
+    fn from_state(state: Arc<State<K>>) -> Self {
+        Self { state: Some(state) }
+    }
+
+    pub fn mark_entry_accessed(&self, key: K) {
+        if let Some(state) = self.state.as_ref() {
+            state.mark_entry_accessed(key);
+        }
+    }
+
+    pub fn mark_entry_removed(&self, key: K) {
+        if let Some(state) = self.state.as_ref() {
+            state.mark_entry_removed(key);
+        }
+    }
+
+    pub fn drain_expired_items(&self, entries: &mut Vec<K>) {
+        if let Some(state) = self.state.as_ref() {
+            let mut inner = state.inner.lock().unwrap();
+
+            // Calculate the cutoff time for entries to be considered expired.
+            //
+            // Our math here shouldn't ever fail, but if it does, we just return and wait for the next attempt.
+            let now = get_unix_timestamp();
+            let last_seen_cutoff = match now.checked_sub(inner.time_to_idle.as_secs()) {
+                Some(cutoff) => cutoff,
+                None => return,
+            };
+
+            let expired_entries = inner
```

Comment:
Should we call `process_pending_ops` before draining? Since wouldn't there be some `ExpirationOp::Accessed` that might be in the queue which would affect the access time and so potentially some entries would be removed even if they were accessed recently?

Reply:
Good catch!

This would indeed tighten up how precise we are with expirations. I'll fix this up.